### PR TITLE
Fix c++11 0.27

### PR DESCRIPTION
**Pull request by clanmills**
This is a "quick hack" to provide support for C++11 in Exiv2 v0.27.3.  If the review is favourable, this can go into v0.27.3 RC2.

A "proper" fix is to port #601 to `0.27-maintenance` and I'm willing to do that if we ever do Exiv2 v0.27.4.  I'm unwilling to make such a major change to the code for 0.27.3.

I don't know if we need to do further "dots".  The API for v0.28 is different and applications will have to adjust to build with v0.28.  And we have removed features (video, ssh and eps) from 0.28.

I feel the way forward is to focus on getting 0.28 published later in 2020.

By the way, there are changes in this PR relating to samples/geotag.cpp.  Those are irrelevant and should be on another branch.  Git strikes again.

### CODE REVIEW

File: cmake/compilerFlags.cmake [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # These flags applies to exiv2lib, the applications, and to the xmp code
 
+if ( EXIV2_BUILD_USE_C++11 )
```
Comment (piponazo) [inline cmake/compilerFlags.cmake:3]: I think we wound need to following. Replace these lines by:

```
if ( EXIV2_BUILD_USE_C++11 ) 
  set(CMAKE_CXX_STANDARD 11)
else()
  set(CMAKE_CXX_STANDARD 98)
endif()

set(CMAKE_CXX_STANDARD_REQUIRED ON)	
set(CMAKE_CXX_EXTENSIONS ON)
```

And more important is to remove all the usages of:

```
target_compile_features(XXX PRIVATE cxx_std_98)
```

File: cmake/compilerFlags.cmake [hunk 2]
```diff
@@ -90,6 +89,10 @@ if ( MINGW OR UNIX OR MSYS ) # MINGW, Linux, APPLE, CYGWIN
             endif()
         endif()
     endif()
+    if ( EXIV2_BUILD_USE_C++11 ) 
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-deprecated")
```
Comment (piponazo) [inline cmake/compilerFlags.cmake:93]: Please, replace the usage of `CMAKE_CXX_FLAGS` by the more appropriate usage of `add_compile_options`:

```
  add_compile_options(-Wno-deprecated)
```

File: include/exiv2/config.h [hunk 3]
```diff
@@ -93,5 +93,15 @@ typedef int pid_t;
...
 //////////////////////////////////////
 
+// https://softwareengineering.stackexchange.com/questions/291141/how-to-handle-design-changes-for-auto-ptr-deprecation-in-c11
+#if __cplusplus >= 201103L
+  #include <memory>
+  #include <sys/types.h>
+  #include <unistd.h>
+  template <typename T>
+  using auto_ptr = std::unique_ptr<T>;
+#else
```
Comment (piponazo) [inline include/exiv2/config.h:103]: That's a clever trick, I did not think about that. 

I think you can remove lines 103 & 104 from here. We do not need the `using` instruction when not using C++11.

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -91,10 +91,6 @@ generate_export_header(exiv2lib
     STATIC_DEFINE exiv2lib_STATIC
 )
 
-target_compile_features(exiv2lib_int PRIVATE cxx_std_98)
```
Comment (piponazo) [inline src/CMakeLists.txt:94]: There are some other usages of `target_compile_features` in the unitTests and samples folders.

**Review by piponazo — CHANGES_REQUESTED** [review]

**Comment by clanmills** [comment]
Thanks for the review @piponazo   I believe I've removed every `target_compile_features(... cxx_std_98)`
```
1100 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ find . -name *.cmake | xargs grep 98
1101 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ find . -name CMakeLists.txt | xargs grep 98
1102 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ find . -name *.cmake | xargs grep target_compile_features
1103 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ find . -name CMakeLists.txt | xargs grep target_compile_features
1104 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ 
```
I now understand what Mr Bruins was shouting about.  We shouldn't try to dictate the tools.  We only have to document how to use them!  So we don't need any special magic about USE_C++11.  The correct way to do this is documented in README.md as:

<div id="2-17">

### 2.17 Building with C++11 and other compilers

Exiv2 uses the default compiler for your system.  Exiv2 v0.27 was written to the C++ 1998 standard and will compile with C++11 and C++14.  As _auto\_ptr support is no longer in C++17, you cannot build with that system._  Exiv2 v0.28 and later do not use auto\_ptr and are fully supported with all compilers.

To generate a build with C++11:

```bash
$ cd <exiv2dir>
$ mkdir build ; cd build
$ cmake .. -DCMAKE_CXX_STANDARD=11 -DCMAKE_CXX_FLAGS=-Wno-deprecated
$ make
```

The option -DCMAKE\_CXX\_FLAGS=-Wno-deprecated suppresses warnings from C++11 concerning auto\_ptr which is deprecated.

**Comment by clanmills** [comment]
This has been very useful.  It shines a clear light on our release strategy from here.

We should only do maintenance on 0.27-maintenance.  So, we leave the auto_ptr alone.  It is what it is.

We should port new features/fixes in 0.27-maintenance into `master` and publish 0.28 later this year.  I'll make the release and do the publishing on exiv2.org.

How does that sound to you?

File: cmake/compilerFlags.cmake (hunk 2 above)
Comment (clanmills) [reply cmake/compilerFlags.cmake:93]: I've removed that 100%.  I don't mess with the compile_options for `-Wno-deprecated`  The user can pass that from the command when required.  And that's what I've documented in README.md.

File: include/exiv2/config.h (hunk 3 above)
Comment (clanmills) [reply include/exiv2/config.h:103]: I can't remember what `using` does.  I got the code from StackOverflow (in the comment).

File: src/CMakeLists.txt (hunk 4 above)
Comment (clanmills) [reply src/CMakeLists.txt:94]: I've got rid of all the `target_compile_features` magic.

File: README.md [hunk 5]
```diff
@@ -696,6 +697,24 @@ You will find that 3 tests fail at the end of the test suite.  It is safe to ign
...
+### 2.17 Building with C++11 and other compilers
+
+Exiv2 uses the default compiler for your system.  Exiv2 v0.27 was written to the C++ 1998 standard and will compile with C++11 and C++14.  As auto_ptr support is no longer in C++17, you cannot build with that system.  Exiv2 v0.28 and later do not use auto_ptr and are fully supported with all compilers.
+
+To generate a build with C++11:
+
+```bash
+$ cd <exiv2dir>
+$ mkdir build ; cd build
+$ cmake .. -DCMAKE_CXX_STANDARD=11 -DCMAKE_CXX_FLAGS=-Wno-deprecated
```
Comment (piponazo) [inline README.md:711]: Great! This is exactly how it should be used. I did not think about it before.

**Comment by piponazo** [comment]
Ey @clanmills this PR broke the CI in most of the services, let me know if I can help you with this.

**Comment by clanmills** [comment]
The CI is broken because the output log is so long with all the deprecation warnings.  We should probably update the CI to use `-Wno_deprecated`

I removed the geotag related stuff from the PR, and it removed the files from the 0.27-maintenance branch.

Let me get the product back together and I'll build it on the Mac Mini (macOS, Ubuntu, Cygwin, MinGW, VS2019, SunOS, FreeBSD and NetBSD).   Then I'll relax.

**Pull request merged**